images/sds-health-watcher-controller/src/pkg/controller/block_device_labels_watcher.go

```diff
@@ -0,0 +1,182 @@
+			shouldRetry = true
+			continue
+		}
+
+		log.Debug(fmt.Sprintf("[ReconcileBlockDeviceLabels] tries to configure a selector from blockDeviceSelector of the LVMVolumeGroup %s", lvg.Name))
+		selector, err := metav1.LabelSelectorAsSelector(lvg.Spec.BlockDeviceSelector)
+		if err != nil {
+			return false, err
+		}
+		log.Debug(fmt.Sprintf("[ReconcileBlockDeviceLabels] successfully configured a selector from blockDeviceSelector of the LVMVolumeGroup %s", lvg.Name))
+
+		usedBdNames := make(map[string]struct{}, len(lvg.Status.Nodes[0].Devices))
+		for _, n := range lvg.Status.Nodes {
+			for _, d := range n.Devices {
+				usedBdNames[d.BlockDevice] = struct{}{}
+			}
+		}
+		if lvg.Labels == nil {
+			lvg.Labels = make(map[string]string)
+		}
+
+		shouldTrigger := false
+		switch selector.Matches(labels.Set(blockDevice.Labels)) {
+		case true:
+			log.Debug(fmt.Sprintf("[ReconcileBlockDeviceLabels] BlockDevice %s matches a blockDeviceSelector of the LVMVolumeGroup %s", blockDevice.Name, lvg.Name))
+			if _, used := usedBdNames[blockDevice.Name]; !used {
+				log.Info(fmt.Sprintf("[ReconcileBlockDeviceLabels] the BlockDevice %s matches the LVMVolumeGroup %s blockDeviceSelector, but is not used yet. Add the label %s to provide LVMVolumeGroup resource configuration update", blockDevice.Name, lvg.Name, LVGUpdateTriggerLabel))
+				shouldTrigger = true
+				break
+			}
+
+			// for the case when BlockDevice stopped match the LVG blockDeviceSelector and then start again
+			for _, c := range lvg.Status.Conditions {
+				if c.Type == internal.TypeVGConfigurationApplied && c.Status == metav1.ConditionFalse {
+					log.Warning(fmt.Sprintf("[ReconcileBlockDeviceLabels] the BlockDevice %s matches the LVMVolumeGroup %s blockDeviceSelector, but the LVMVolumeGroup has condition %s in status False. Add the label %s to provide LVMVolumeGroup resource configuration update", blockDevice.Name, lvg.Name, c.Type, LVGUpdateTriggerLabel))
+					shouldTrigger = true
+					break
+				}
+			}
+
+			log.Debug(fmt.Sprintf("[ReconcileBlockDeviceLabels] the BlockDevice %s matches the LVMVolumeGroup %s blockDeviceSelector and already used by the resource", blockDevice.Name, lvg.Name))
+		case false:
+			log.Debug(fmt.Sprintf("[ReconcileBlockDeviceLabels] BlockDevice %s does not match a blockDeviceSelector of the LVMVolumeGroup %s", blockDevice.Name, lvg.Name))
+			if _, used := usedBdNames[blockDevice.Name]; used {
+				log.Warning(fmt.Sprintf("[ReconcileBlockDeviceLabels] the BlockDevice %s does not match the LVMVolumeGroup %s blockDeviceSelector, but is used by the resource. Add the label %s to provide LVMVolumeGroup resource configuration update", blockDevice.Name, lvg.Name, LVGUpdateTriggerLabel))
+				shouldTrigger = true
+				break
+			}
+
+			log.Debug(fmt.Sprintf("[ReconcileBlockDeviceLabels] the BlockDevice %s does not match the LVMVolumeGroup %s blockDeviceSelector and is not used by the resource", blockDevice.Name, lvg.Name))
+		}
```

Comment:
Extracting this block into a function `shouldTrigger(...)` could be beneficial for readability. Also it will remove a few lines, since you can always just `return true` from there